src/ProductConstructionService/ProductConstructionService.ReproTool/ReproTool.cs

```diff
@@ -0,0 +1,280 @@
+        await UpdateRemoteFileAsync(newBranchName, productRepoUri, productRepoForkUri, SourceManifestPath);
+
+        return AsyncDisposableValue.Create(newBranchName, async () =>
+        {
+            logger.LogInformation("Cleaning up temporary branch {branchName}", newBranchName);
+            try
+            {
+                await DeleteGitHubBranchAsync(VmrForkRepoName, newBranchName);
+            }
+            catch (Exception)
+            {
+                // If this throws an exception the most likely cause is that the branch was already deleted
+            }
+        });
+    }
+
+    private async Task DeleteGitHubBranchAsync(string repo, string branch) => await ghClient.Git.Reference.Delete(MaestroAuthTestOrgName, repo, $"heads/{branch}");
+
+    private async Task UpdateRemoteFileAsync(string branch, string productRepoUri, string productRepoForkUri, string filePath)
+    {
+        // Fetch remote file and replace the product repo URI with the repo we're testing on        
+        var sourceMappingsFile = (await ghClient.Repository.Content.GetAllContentsByRef(
+            MaestroAuthTestOrgName,
+            VmrForkRepoName,
+            filePath,
+            branch)).FirstOrDefault() ??
+                throw new Exception($"Failed to find file {SourceMappingsPath} in {MaestroAuthTestOrgName}" +
+                    $"/{VmrForkRepoName} on branch {SourceMappingsPath}");
+
+        // Replace the product repo uri with the forked one
+        var updatedSourceMappings = sourceMappingsFile.Content.Replace(productRepoUri, productRepoForkUri);
+        UpdateFileRequest update = new(
+            $"Update {productRepoUri} source mapping",
+            updatedSourceMappings,
+            sourceMappingsFile.Sha,
+            branch);
+
+        await ghClient.Repository.Content.UpdateFile(
+            MaestroAuthTestOrgName,
+            VmrForkRepoName,
+            filePath,
+            update);
+    }
+
+    private async Task PrepareProductRepoForkAsync(
+        string productRepoUri,
+        string productRepoForkUri,
+        string productRepoBranch)
+    {
+        (var org, var name) = GetRepoOwnerAndName(productRepoUri);
+        // Check if the product repo fork already exists
+        var allRepos = await ghClient.Repository.GetAllForOrg(MaestroAuthTestOrgName);
+        
+        // If we already have a fork in maestro-auth-test, sync the branch we need with the source
+        if (allRepos.FirstOrDefault(repo => repo.HtmlUrl == productRepoForkUri) != null)
+        {
+            await SyncForkAsync(org, name, productRepoBranch);
+        }
+        // If we don't, create a fork
+        else
+        {
+            await ghClient.Repository.Forks.Create(org, name, new NewRepositoryFork { Organization = MaestroAuthTestOrgName });
+        }
+        await AddRepositoryToBarIfMissingAsync(productRepoForkUri);
```

Comment:
Can we put in some logging so that we know what the tool is doing?

Reply:
added